# Day 3: Binary Diagnostic 

## Part one

The submarine has been making some odd creaking noises, so you ask it to produce a diagnostic report just in case.

The diagnostic report (your puzzle input) consists of a list of binary numbers which, when decoded properly, can tell you many useful things about the conditions of the submarine. The first parameter to check is the **power consumption**.

You need to use the binary numbers in the diagnostic report to generate two new binary numbers (called the **gamma rate** and the **epsilon rate**). The power consumption can then be found by multiplying the gamma rate by the epsilon rate.

Each bit in the gamma rate can be determined by finding the **most common bit in the corresponding position** of all numbers in the diagnostic report. For example, given the following diagnostic report:

```
00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010
```

Considering only the first bit of each number, there are five `0` bits and seven `1` bits. Since the most common bit is `1`, the first bit of the gamma rate is `1`.

The most common second bit of the numbers in the diagnostic report is `0`, so the second bit of the gamma rate is `0`.

The most common value of the third, fourth, and fifth bits are `1`, `1`, and `0`, respectively, and so the final three bits of the gamma rate are `110`.

So, the gamma rate is the binary number `10110`, or `22` in decimal.

The epsilon rate is calculated in a similar way; rather than use the most common bit, the least common bit from each position is used. So, the epsilon rate is `01001`, or `9` in decimal. Multiplying the gamma rate (`22`) by the epsilon rate (`9`) produces the power consumption, `198`.

Use the binary numbers in your diagnostic report to calculate the gamma rate and epsilon rate, then multiply them together. **What is the power consumption of the submarine?** (Be sure to represent your answer in decimal, not binary.)

In [67]:
input_data = []

with open('data/input.txt', 'r') as f:
    for line in f.read().splitlines():
        input_data.append(line)
        
input_data[:10]

['011111111100',
 '100001011111',
 '011010010010',
 '100110110110',
 '001000110001',
 '110010001010',
 '110000010010',
 '010110100110',
 '000101000110',
 '100101010010']

In [25]:
most_common_bits = []

for i in range(0, len(input_data[0])):
    most_common_bits.append({"num_of_0": 0, "num_of_1": 0, "most_common": '0', "least_common": '1', })

most_common_bits

for line in input_data:
    bit_number = 0
    for bit in line:
        if '0' == bit:
            most_common_bits[bit_number]['num_of_0'] += 1
        else:
            most_common_bits[bit_number]['num_of_1'] += 1
        bit_number += 1
        
for bit_detail in most_common_bits:
    if bit_detail['num_of_1'] > bit_detail['num_of_0']:
        bit_detail['most_common'] = '1'
        bit_detail['least_common'] = '0'

most_common_bits

[{'num_of_0': 502, 'num_of_1': 498, 'most_common': '0', 'least_common': '1'},
 {'num_of_0': 501, 'num_of_1': 499, 'most_common': '0', 'least_common': '1'},
 {'num_of_0': 462, 'num_of_1': 538, 'most_common': '1', 'least_common': '0'},
 {'num_of_0': 519, 'num_of_1': 481, 'most_common': '0', 'least_common': '1'},
 {'num_of_0': 499, 'num_of_1': 501, 'most_common': '1', 'least_common': '0'},
 {'num_of_0': 508, 'num_of_1': 492, 'most_common': '0', 'least_common': '1'},
 {'num_of_0': 510, 'num_of_1': 490, 'most_common': '0', 'least_common': '1'},
 {'num_of_0': 506, 'num_of_1': 494, 'most_common': '0', 'least_common': '1'},
 {'num_of_0': 480, 'num_of_1': 520, 'most_common': '1', 'least_common': '0'},
 {'num_of_0': 497, 'num_of_1': 503, 'most_common': '1', 'least_common': '0'},
 {'num_of_0': 483, 'num_of_1': 517, 'most_common': '1', 'least_common': '0'},
 {'num_of_0': 501, 'num_of_1': 499, 'most_common': '0', 'least_common': '1'}]

In [28]:
gamma_rate = ''
epsilon_rate = ''

bit_number = 0
for bit_detail in most_common_bits:
    gamma_rate += bit_detail['most_common']
    epsilon_rate += bit_detail['least_common']
    
    bit_number += 1
    
print('gamma_rate: \'%s\' - epsilon_rate: \'%s\'' % (gamma_rate, epsilon_rate, ))

gamma_rate: '001010001110' - epsilon_rate: '110101110001'


In [31]:
int(gamma_rate, 2) * int(epsilon_rate, 2)

2250414

Your puzzle answer was `2250414`.

## Part two

Next, you should verify the **life support rating**, which can be determined by multiplying the **oxygen generator rating** by the **CO2 scrubber rating**.

Both the oxygen generator rating and the CO2 scrubber rating are values that can be found in your diagnostic report - finding them is the tricky part. Both values are located using a similar process that involves filtering out values until only one remains. Before searching for either rating value, start with the full list of binary numbers from your diagnostic report and **consider just the first bit** of those numbers. Then:

- Keep only numbers selected by the **bit criteria** for the type of rating value for which you are searching. Discard numbers which do not match the bit criteria.
- If you only have one number left, stop; this is the rating value for which you are searching.
- Otherwise, repeat the process, considering the next bit to the right.

The **bit criteria** depends on which type of rating value you want to find:

- To find **oxygen generator rating**, determine the **most common** value (0 or 1) in the current bit position, and keep only numbers with that bit in that position. If `0` and `1` are equally common, keep values with a 1 in the position being considered.
- To find **CO2 scrubber rating**, determine the **least common** value (0 or 1) in the current bit position, and keep only numbers with that bit in that position. If `0` and `1` are equally common, keep values with a 0 in the position being considered.

For example, to determine the oxygen generator rating value using the same example diagnostic report from above:

- Start with all 12 numbers and consider only the first bit of each number. There are more 1 bits (`7`) than 0 bits (`5`), so keep only the 7 numbers with a 1 in the first position: `11110`, `10110`, `10111`, `10101`, `11100`, `10000` and `11001`.
- Then, consider the second bit of the 7 remaining numbers: there are more 0 bits (`4`) than 1 bits (`3`), so keep only the 4 numbers with a `0` in the second position: `10110`, `10111`, `10101` and `10000`.
- In the third position, three of the four numbers have a `1`, so keep those three: `10110`, `10111`, and `10101`.
- In the fourth position, two of the three numbers have a `1`, so keep those two: `10110` and `10111`.
- In the fifth position, there are an equal number of `0` bits and `1` bits (one each). So, to find the **oxygen generator rating**, keep the number with a `1` in that position: `10111`.
- As there is only one number left, stop; the **oxygen generator rating** is `10111`, or `23` in decimal.

Then, to determine the CO2 scrubber rating value from the same example above:

- Start again with all 12 numbers and consider only the first bit of each number. There are fewer 0 bits (`5`) than 1 bits (`7`), so keep only the 5 numbers with a `0` in the first position: `00100`, `01111`, `00111`, `00010`, and `01010`.
- Then, consider the second bit of the 5 remaining numbers: there are fewer 1 bits (`2`) than 0 bits (`3`), so keep only the 2 numbers with a `1` in the second position: `01111` and `01010`.
- In the third position, there are an equal number of 0 bits and 1 bits (one each). So, to find the **CO2 scrubber rating**, keep the number with a `0` in that position: `01010`.
- As there is only one number left, stop; the **CO2 scrubber rating** is `01010`, or `10` in decimal.

Finally, to find the life support rating, multiply the oxygen generator rating (`23`) by the CO2 scrubber rating (`10`) to get `230`.

Use the binary numbers in your diagnostic report to calculate the oxygen generator rating and CO2 scrubber rating, then multiply them together. **What is the life support rating of the submarine?** (Be sure to represent your answer in decimal, not binary.)

In [68]:
def ratings_process(ratings, bit_position, selected_ratings=[]):
    processed_ratings = {"num_of_0": 0, 
                        "num_of_1": 0, 
                        "most_common": '1', 
                        "least_common": '0', 
                        "ratings_0": [], 
                        "ratings_1": []}
    
    if not selected_ratings:
        range_max = len(ratings)
    else:
        range_max = len(selected_ratings)

    for i in range(0, range_max):
        if not selected_ratings:
            rating_number = i
        else:
            rating_number = selected_ratings[i]
            
        if "0" == ratings[rating_number][bit_position]:
            processed_ratings["num_of_0"] += 1
            processed_ratings["ratings_0"].append(rating_number)
        else:
            processed_ratings["num_of_1"] += 1
            processed_ratings["ratings_1"].append(rating_number)
            
    if processed_ratings["num_of_0"] > processed_ratings["num_of_1"]:
        processed_ratings["most_common"] = '0'
        processed_ratings["least_common"] = '1'
            
    return processed_ratings

oxygen_rate = ''
selected_ratings = []
for bit_position in range(0, len(input_data[0])):
    processed_ratings = ratings_process(input_data, bit_position, selected_ratings=selected_ratings)
    
    if '0' == processed_ratings['most_common']:
        selected_ratings = processed_ratings['ratings_0']
    else:
        selected_ratings = processed_ratings['ratings_1']
        
    if len(selected_ratings) == 1:
        oxygen_rate = input_data[selected_ratings[0]]
        
co2_rate = ''
selected_ratings = []
for bit_position in range(0, len(input_data[0])):
    processed_ratings = ratings_process(input_data, bit_position, selected_ratings=selected_ratings)
    
    if '0' == processed_ratings['least_common']:
        selected_ratings = processed_ratings['ratings_0']
    else:
        selected_ratings = processed_ratings['ratings_1']
        
    if len(selected_ratings) == 1:
        co2_rate = input_data[selected_ratings[0]]
        
print("oxygen_rate: %s - co2_rate: %s" % (oxygen_rate, co2_rate, ))

oxygen_rate: 011110001111 - co2_rate: 110001001001


In [69]:
int(oxygen_rate, 2) * int(co2_rate, 2)

6085575

Your puzzle answer was `6085575`.